### ISSUE TITLE
Purpose of `pending_writes`?

### ISSUE THREAD

**Issue by iwasaki-kenta** [nomic-io/merk#29]
Is there any reason for having `Link::Modified`? Apart from enforcing some invariants like preventing modified trees from being pruned, I suppose removing `Link::Modified` would save a bit of ops/sec.

Apart from that, I am also curious if there is a need for `pending_writes` in `Link::Modified`.

**Comment by mappum** [comment]
Hey @iwasaki-kenta, thanks for your interest in merk.

> Is there any reason for having Link::Modified?

Maybe it should be better documented but it's there because applying changes to the tree actually takes 2 passes: a first pass where we just change the tree structure (marking some links as `Modified`, where their hash has not been computed yet) and a second pass where we recompute hashes and write the changes to RocksDB.

I went with this strategy because sometimes we traverse through subtrees more than once, for instance when we delete a node and all of its ancestors now need to be re-hashed, so if we computed hashes as we traversed then sometimes we would rehash a node multiple times in one batch.

Another reason is that in the future maybe there will be applications which apply changes to the tree multiple times before re-hashing and committing to disk (e.g. they could do the "first pass" N times, then the "second pass" once they need the new root hash).

I actually originally implemented it without `Link::Modified` and just recomputed hashes as the tree was traversed, and there wasn't really a significant performance difference between the two strategies.

>Apart from that, I am also curious if there is a need for pending_writes in Link::Modified.

This isn't currently used, but it is there for the future addition of concurrent ops, where we will be able to use `pending_writes` as a heuristic for evenly breaking up the work among workers for maximum CPU utilization.
